### PR TITLE
(Feat) Posting Journal Editing

### PR DESCRIPTION
**Pull request by jniles**
This PR adds the basic format for creating ui-grid editors by implementing a datepicker editor.  The editor is found in the directive `bhGridDateEditor.js`.  For technical details, see the code!

**New Methods**
Additionally, this commit introduces two new functions: `util.before()` and `util.after()` for calling functions before or after specific class/object method calls.  They can be used like this:

``` js
var klass = { method : p => console.log('Called with p:', p) };
util.before(klass, 'method', p => {
  console.log('I got injected before!  I found parameter:', p);
});

klass.method(2);
// => I got injected before!  I found parameter: 2
// => Called with p: 2

```

`util.after()` is similar, except the call is made after the method call is executed.

**New Features**
This PR introduces a loading indicator to the Posting Journal.  It leverages the `bhGridLoadingIndicator` directive.

**Tests**
This PR also introduces unit tests for:
1. `util.once()` introduced in #598.
2. `util.before()` introduced here.
3. `util.after()` introduced here.

**TODO**
This PR only implements an editor - none of the editing logic.  Here was what remains to be done:
1. ~~Limit edits to a single transaction at a time.~~
2. ~~Implement a "Save" button to validate and terminate the edits on a transaction.~~
3. When trying to save an `$invalid` input value either:
   1. Replace the edits with the original data.
   2. Do not allow the user to quit editing until the inputs are `$valid`
4. ~~Highlight the transaction currently being edited.~~

---

Thank you for contributing!

Before submitting this pull request, please verify that you have:
- [x] Run your code through JSHint.  [Check out our styleguide](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/docs/STYLEGUIDE.md).
- [x] Run integration tests.
- [x] Run end-to-end tests.
- [x] Accurately described the changes your are making in this pull request.

For a more detailed checklist, [see the official review checklist](https://docs.google.com/document/d/1nupLVLRXgSZJQo_acLgrwvPnN8RukfSiwRhSToj81uU/pub) that this PR will be evaluated against.

Ensuring that the above checkboxes are completed will help speed the review process and help build a stronger application.  Thanks!

### CODE REVIEW

**Comment by jniles** [comment]
@sfount, could I get a review?

**Comment by jniles** [comment]
@sfount, can I get a review?

File: client/src/partials/journal/journal.html [hunk 1]
```diff
@@ -2,7 +2,21 @@
   <div class="bhima-title">
     <ol class="headercrumb">
       <li class="static">{{ "TREE.FINANCE" | translate }}</li>
-      <li class="title">{{ "POSTING_JOURNAL.TITLE" | translate }}</li>
+      <li class="title">
+        {{ "POSTING_JOURNAL.TITLE" | translate }}
+
+        <span class="badge badge-primary text-uppercase" ng-show="JournalCtrl.transactions.isEditing()">
+          <i class="fa fa-street-view"></i>
+          Edit Mode
```
Comment (sfount) [inline client/src/partials/journal/journal.html]: This should eventually be translated - the feature is still in early development so this is just a reminder comment.

File: client/src/partials/journal/journal.html [hunk 2]
```diff
@@ -2,7 +2,21 @@
...
+        </span>
+
+        <span class="badge badge-warning text-uppercase" ng-show="JournalCtrl.transactions._highlights.length > 0">
+          {{ JournalCtrl.transactions._highlights.length }}
+          <span ng-if="JournalCtrl.transactions._highlights.length == 1">Highlight</span>
```
Comment (sfount) [inline client/src/partials/journal/journal.html]: Are highlights a feature that will be implemented in the future or that were used during development/ debugging? 
I can't see any way that this state is ever reached in this PR.

Comment (jniles) [inline client/src/partials/journal/journal.html]: They were mostly for prototyping, although they may be useful in highlighting transactions is coming from a previous state.  Something like:

`#/journal/:uuid` --> highlight transaction with record UUID `:uuid`.  This is unimplemented.

But I'll take this out for the time being.  They are easy to add back in.

File: client/src/js/services/grid/GridSorting.js [hunk 3]
```diff
@@ -77,23 +79,14 @@ function GridSortingService() {
...
-    var cacheGridApi = gridOptions.onRegisterApi;
-
-    // Global sorting configuration
+    // global sorting configuration
     gridOptions.enableSorting = true;
 
-    // // Register for the Grid API
-    gridOptions.onRegisterApi = function onRegisterApi(api) {
+    // register for the grid API
+    util.after(gridOptions, 'onRegisterApi', function onRegisterApi(api) {
```
Comment (sfount) [inline client/src/js/services/grid/GridSorting.js:86]: Cool 👍

**Comment by sfount** [comment]
@jniles This is great base for such a complex feature, really well designed and thought through. 

This sets the journal up for connecting the edit functionality with the server API writing to the database. The one thing I noticed through testing the feature is in line with the TODOs you have in the description. On editing a currency value any kind of value can be entered (string for example) and this can then be saved - because the currency filter does not understand strings nothing is displayed in the cell even though the underlying value has been changed to the new string; if it's possible I would recommend the next iteration blocks leaving the edit state until it is either cancelled or $valid. 

Really great work, code LGTM 👍

**Comment by jniles** [comment]
@sfount, thanks for the review.

I'll work on saving `$invalid` form elements and what the UX should be like.  Once that lands, I'll give you an update here for another review!

**Comment by jniles** [comment]
@sfount, I've removed highlighting and added i18n keys.

For the `$invalid` state, I think a new PR should be submitted to untangle the mechanics of editing from the correctness of validation.  I'll work on that in a separate branch.  Could I get another review of this one?

**Comment by sfount** [comment]
Changes LGTM.

**Pull request merged**